### PR TITLE
Revise mutable_arithmetics.jl

### LINKED ISSUE #61: Parameter JuMP detects QuadraticExpressions with GAEP operations.
When performing operations with parametrized affine expressions, the behavior of ParameterJuMP is inconsistent with JuMP. Here is a MWE of what I am observing while trying to reproduce the issue in #57 

```Julia
using JuMP
using Random
m = Model() 
I = 1:1000
J = 1:1000
vals = rand(1000)
@variable(m, x[I, J] >= 0)
exps = @expression(m, additions[ix = I], sum(x[ix,:]) - vals[ix])
@constraint(m, con, sum(exps) >= 0);
typeof(con)
ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}
```

```Julia
using ParameterJuMP
using Random
mwp = ModelWithParams()
I = 1:1000
J = 1:1000
vals = rand(1000)
param_vals = add_parameters(mwp, vals)
@variable(mwp, x[I, J] >= 0)
expswp = @expression(mwp, additions[ix = I], sum(x[ix,:]) - param_vals[ix])
@constraint(mwp, con, sum(expswp) >= 0);

Collection of expressions with @expression must be linear. For quadratic expressions, use your own array.

Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] macro expansion at /Users/jdlara/.julia/packages/JuMP/ibcEh/src/macros.jl:1068 [inlined]
 [3] top-level scope at ./In[35]:7
```

### LINKED ISSUE #62: Sum of Array{T} where T is GAEv, GAEp or PAE inside macro returns error
Hi,

I wanted to define a linear program with parameters using matrix and I observed this undesirable behavior when we use double affine expressions in matrix form with `@constraint` and `@expression` macros. I use ParameterJuMP v0.1.1.
```julia
using ParameterJuMP, JuMP

m = ModelWithParams()

@variable(m, x[1:2])
θ = add_parameters(m, 2)
A = ones(2,2)

@constraint(m, A*x + A*θ  .<= 0)  # Returns following error
@expression(m, A*x + A*θ)   # Returns following error as well

expr = A*x + A*θ  # Works
@constraint(m, expr  .<= 0) # Works
```

The error observed with `@macro` and `@expression` macros:
```julia
ERROR: LoadError: MethodError: no method matching add_to_expression!(::GenericAffExpr{Float64,VariableRef}, ::GenericAffExpr{Float64,ParameterRef})
```

Is it because we should not replace a GAEv by a DGAE so you did not overload `add_to_expression!` for this case while you did it for `+`?

### LINKED ISSUE #66: ParameterJuMP.DoubleGenericAffExpr alegra failing in @constraint call
When adding a constraint with the expression declaration inside, mutable arithmetics doesn't seem to be able to perform the operation

```Julia
using JuMP
using ParameterJuMP
m = ModelWithParams()
x = @variable(m)
y = add_parameter(m, 10.0)
@constraint(m, x == (x+y)*10)
z = @variable(m)
@constraint(m, x == (x+y+z)*10)
@constraint(m, x == (x+y*5+z)*10)
```

Returns this error 
```
ERROR: ArgumentError: Cannot call `mutable_operate!(sub_mul, ::ParameterJuMP.DoubleGenericAffExpr{Float64,VariableRef,ParameterRef}, ::ParameterRef, ::Int64, ::Int64)` as objects of type `ParameterJuMP.DoubleGenericAffExpr{Float64,VariableRef,ParameterRef}` cannot be modifed to equal the result of the operation. Use `operate!` instead which returns the value of the result (possibly modifying the first argument) to write generic code that also works when the type cannot be modified.
Stacktrace:
 [1] mutable_operate_fallback(::MutableArithmetics.NotMutable, ::Function, ::ParameterJuMP.DoubleGenericAffExpr{Float64,VariableRef,ParameterRef}, ::ParameterRef, ::Int64, ::Vararg{Int64,N} where N) at /Users/jdlara/.julia/packages/MutableArithmetics/hS4h3/src/interface.jl:203
 [2] mutable_operate!(::Function, ::ParameterJuMP.DoubleGenericAffExpr{Float64,VariableRef,ParameterRef}, ::ParameterRef, ::Int64, ::Int64) at /Users/jdlara/.julia/packages/MutableArithmetics/hS4h3/src/interface.jl:221
 [3] sub_mul(::GenericAffExpr{Float64,VariableRef}, ::ParameterRef, ::Int64, ::Int64) at /Users/jdlara/.julia/packages/JuMP/CZ8vV/src/mutable_arithmetics.jl:196
 [4] operate(::typeof(MutableArithmetics.sub_mul), ::GenericAffExpr{Float64,VariableRef}, ::ParameterRef, ::Int64, ::Int64) at /Users/jdlara/.julia/packages/MutableArithmetics/hS4h3/src/interface.jl:88
 [5] operate_fallback!(::MutableArithmetics.NotMutable, ::Function, ::GenericAffExpr{Float64,VariableRef}, ::ParameterRef, ::Int64, ::Int64) at /Users/jdlara/.julia/packages/MutableArithmetics/hS4h3/src/interface.jl:275
 [6] operate!(::typeof(MutableArithmetics.sub_mul), ::GenericAffExpr{Float64,VariableRef}, ::ParameterRef, ::Int64, ::Int64) at /Users/jdlara/.julia/packages/MutableArithmetics/hS4h3/src/rewrite.jl:67
 [7] top-level scope at /Users/jdlara/.julia/packages/MutableArithmetics/hS4h3/src/rewrite.jl:224
 [8] top-level scope at /Users/jdlara/.julia/packages/JuMP/CZ8vV/src/macros.jl:380
```

If alternatively, I try to use the expression as follows: 

```Julia
using JuMP
using ParameterJuMP
m = ModelWithParams()
x = @variable(m)
y = add_parameter(m, 10.0)
@constraint(m, x == (x+y)*10)
z = @variable(m)
@constraint(m, x == (x+y+z)*10)
exp = (x+y*5+z)*10
@constraint(m, exp == 0)
```

Returns this error 

```
ERROR: MethodError: no method matching shift_constant(::MathOptInterface.EqualTo{Float64}, ::Int64)
Closest candidates are:
  shift_constant(::Union{MathOptInterface.EqualTo{T}, MathOptInterface.GreaterThan{T}, MathOptInterface.LessThan{T}}, ::T) where T at /Users/jdlara/.julia/packages/MathOptInterface/DmQBj/src/Utilities/sets.jl:17
  shift_constant(::MathOptInterface.Interval, ::Any) at /Users/jdlara/.julia/packages/MathOptInterface/DmQBj/src/Utilities/sets.jl:20
  shift_constant(::MathOptInterface.Test.UnknownScalarSet, ::Any) at /Users/jdlara/.julia/packages/MathOptInterface/DmQBj/src/Test/modellike.jl:9
Stacktrace:
 [1] build_constraint(::Function, ::ParameterJuMP.DoubleGenericAffExpr{Int64,VariableRef,ParameterRef}, ::MathOptInterface.EqualTo{Float64}) at /Users/jdlara/.julia/packages/ParameterJuMP/yfi1m/src/constraints.jl:162
 [2] top-level scope at /Users/jdlara/.julia/packages/JuMP/CZ8vV/src/macros.jl:381
```

### LINKED ISSUE #68: missing `add_to_expression!` methods
`add_to_expression(PAE1, PAE2, mult)` not supported. 

X-ref: https://github.com/NREL-SIIP/PowerSimulations.jl/blob/bbf42f39e4e1e10beed235fa0f0af74634903b72/src/network_models/ptdf_model.jl#L38

### PR DESCRIPTION
**Pull request by odow**
Closes #68 
Closes #66 
Closes #62 
Closes #61

### CODE REVIEW

File: src/mutable_arithmetics.jl [hunk 1]
```diff
@@ -1,47 +1,137 @@
...
-    JuMP.add_to_expression!(aff.v, other)
+# n-argument functions
+
+function MA.mutable_operate!(
+    op::MA.AddSubMul,
+    aff::PAE,
+    a::Number,
+    b::Number,
+    rhs::ParameterRef,
+)
```
Comment (joaquimg) [inline src/mutable_arithmetics.jl]: Does it cover all orders of numbers and parameters?

File: src/mutable_arithmetics.jl [hunk 2]
```diff
@@ -1,47 +1,137 @@
...
 end
-function JuMP.add_to_expression!(aff::PAE, new_var::JuMP.VariableRef, new_coef)
-    JuMP.add_to_expression!(aff.v, new_coef, new_var)
+
+function MA.mutable_operate!(
+    op::MA.AddSubMul,
+    aff::PAE,
+    rhs::PAE,
+    args::Number...
+)
```
Comment (joaquimg) [inline src/mutable_arithmetics.jl]: Same here...

**Comment by odow** [comment]
@jd-lara want to try out this branch?

**Review by joaquimg — APPROVED** [review]

**Comment by odow** [comment]
@jd-lara try out `master`. I'll hold off tagging a new release until you've tested.

**Comment by jd-lara** [comment]
@odow I am making test cases to try different things.

**Pull request merged**